File: src/main/java/leetCode/problems/LeetCode84.java
/*
 * @Author: aiwujingxin@gmail.com
 * @Date: 2022-06-08 19:41:54
 * @LastEditTime: 2022-06-23 01:25:12
 * @LastEditors: aiwujingxin@gmail.com
 */
package leetCode.problems;

import java.util.Stack;

public class LeetCode84 {

    public int largestRectangleArea(int[] heights) {
        int N = heights.length;
        int[] left = new int[N];
        int[] right = new int[N];
        Stack<Integer> stack = new Stack<>();
        int maxArea = Integer.MIN_VALUE;

        for (int i = 0; i < N; i++) {
            while (!stack.isEmpty() && heights[i] <= heights[stack.peek()]) {
                stack.pop();
            }
            if (stack.isEmpty()) {
                left[i] = 0;
            } else {
                left[i] = stack.peek() + 1;
            }
            stack.push(i);
        }

        while (!stack.isEmpty()) stack.pop();

        for (int i = N - 1; i >= 0; i--) {
            while (!stack.isEmpty() && heights[i] <= heights[stack.peek()]) {
                stack.pop();
            }
            if (stack.isEmpty()) right[i] = N - 1;
            else right[i] = stack.peek() - 1;
            stack.push(i);
        }

        for (int i = 0; i < N; i++) {
            int area = heights[i] * (right[i] - left[i] + 1);
            maxArea = Math.max(maxArea, area);
        }
        return maxArea;
    }
}

File: src/main/GO/leetcode/lists/hot100/LeetCode84.go
package hot100

func largestRectangleArea(heights []int) int {
	if len(heights) == 0 {
		return 0
	}
	left, right := make([]int, len(heights)), make([]int, len(heights))
	// 注意: 最小栈的大小初始化为0
	leftStack := make([]int, 0)
	rightStack := make([]int, 0)
	// 最小栈
	for i := 0; i < len(heights); i++ {
		for len(leftStack) > 0 && heights[leftStack[len(leftStack)-1]] >= heights[i] {
			leftStack = leftStack[:len(leftStack)-1]
		}
		if len(leftStack) == 0 {
			left[i] = -1
		} else {
			left[i] = leftStack[len(leftStack)-1]
		}
		leftStack = append(leftStack, i)
	}
	for i := len(heights) - 1; i >= 0; i-- {
		for len(rightStack) > 0 && heights[rightStack[len(rightStack)-1]] >= heights[i] {
			rightStack = rightStack[:len(rightStack)-1]
		}
		if len(rightStack) == 0 {
			right[i] = len(heights)
		} else {
			right[i] = rightStack[len(rightStack)-1]
		}
		rightStack = append(rightStack, i)
	}
	res := 0
	for i := 0; i < len(heights); i++ {
		res = Max(res, (right[i]-left[i]-1)*heights[i])
	}
	return res
}
